include/libtorrent/bt_peer_connection.hpp

```diff
@@ -159,7 +159,10 @@ namespace libtorrent
 		void on_receive_impl(std::size_t bytes_transferred);
 
 #if !defined(TORRENT_DISABLE_ENCRYPTION) && !defined(TORRENT_DISABLE_EXTENSIONS)
-		virtual int hit_send_barrier(std::vector<boost::asio::mutable_buffer>& iovec) override;
+		// next_barrier
+		virtual
+		std::tuple<int, std::vector<boost::asio::const_buffer>>
+		hit_send_barrier(aux::array_view<boost::asio::mutable_buffer> iovec) override;
```

Comment:
@ssiloti this is part of the main change here. basically, instead of taking a reference to a std::vector as an in-out parameter, an array_view is used (to allow for flexible data structures being passed in, such as a stack allocated iovec) and the buffers to be prepended are returned as the second return value. This shifts the cost of heap allocation from *every time* to just when there are buffers to prepend.

This change obviously trickles through down to the crypto_plugin interface too.